source/common/upstream/thread_aware_lb_impl.cc

```diff
@@ -157,11 +158,49 @@ ThreadAwareLoadBalancerBase::LoadBalancerImpl::chooseHost(LoadBalancerContext* c
     stats_.lb_healthy_panic_.inc();
   }
 
+
+  /*
+    This will be a config parameter under consistent_hashing_lb_config.
+
+    hash_balance_factor: a percentage of load to bound per upstream server.
+    For eg., with a value of 150, no upstream host will get a load more than
+    1.5 times the average load of any other upstream host.
+
+    Default 0 for being disabled.
+
+  */
+  const uint32_t hash_balance_factor = 125;
+
   HostConstSharedPtr host;
   const uint32_t max_attempts = context ? context->hostSelectionRetryCount() + 1 : 1;
   for (uint32_t i = 0; i < max_attempts; ++i) {
     host = per_priority_state->current_lb_->chooseHost(h, i);
 
+    if (hash_balance_factor > 0) {
+      auto normalized_host_weights = per_priority_state->normalized_host_weights_;
+      const auto& it = std::find_if(normalized_host_weights->begin(), normalized_host_weights->end(),
+                        [host] (const std::pair<HostConstSharedPtr, double>& p) { return p.first == host; } );
+      ASSERT(it != normalized_host_weights->end());
+      const double chosen_host_normalized_weight = it->second;
+
+      /*
+        Consistent Hashing with Bounded Load
+        Ref: https://arxiv.org/abs/1608.01350
+      */
+      uint32_t total_slots = ((host->cluster().stats().upstream_rq_active_.value() + 1) * hash_balance_factor + 99) / 100;
+      uint32_t slots = static_cast<uint32_t>(std::ceil(total_slots * chosen_host_normalized_weight));
+
+      if (slots == 0)
+        slots = 1;
+
+      if (host->stats().rq_active_.value() > slots) {
+        ENVOY_LOG_MISC(debug, "ThreadAwareLoadBalancerBase::LoadBalancerImpl::chooseHost: chosen host {} overloaded; active {} > slots {}",
+                host->hostname(), host->stats().rq_active_.value(), slots);
+        continue;
+      }
+    }
+
```

Comment:
Need some suggestions here.  Ideally, this should walk the ring to find the next host that is not overloaded. The above tries only 'max_attempts' number of hosts, which is not guaranteed to find the eligible host.  One option is to add a method to HashingLoadBalancer interface to provider an iterator to walk the ring of hosts.

Reply:
I don't think this belongs here. ThreadAwareLoadBalancerBase happens to only be used by consistent-hash load balancers at the moment, but it's not a given that every future implementation of ThreadAwareLoadBalancerBase will use consistent hashing.

`ThreadAwareLoadBalancer::HashingLoadBalancer` is where I think it makes sense to implement this feature. Probably you'll need to implement `chooseHost` there and modify the maglev and ringhash balancers to suit the needs of this algorithm.

Keep in mind that the retry policy behavior (the `shouldSelectAnotherHost` call below) needs to be preserved at least in the case where bounded load is not enabled, but preferably in both cases.

Reply:
@zuercher moved this logic in a class of its own - please see BoundedLoadHashingLoadBalancer. This doesnt touch HashingLoadBalancer or the classes that implement it, so there is no change in API. The new class kicks in only when the user configures the hash_balance_factor. Please take a look and let me know your comments.